### PR TITLE
Support for Smithy 2

### PR DESCRIPTION
**Pull request by benoitlouy**
This change updates the grammar to support Smithy 2 language specification as described [here](https://awslabs.github.io/smithy/2.0/spec/idl.html#smithy-idl-abnf)

Some of the rules are more permissive than in the official grammar.

Note: I haven't updated the generated source files located in `src/` to keep the PR legible.

### CODE REVIEW

**Comment by keynmol** [comment]
Thanks for the PR!

However, there seems to be a lot of things going on:

1. A lot of reductions are renamed (for no apparent reason)
2. Order of reductions is changed
3. Lots of formatting changes, making it harder to review 

I looked at the PR several times and it looks like a complete rewrite of the grammar (even if it isn't), which I IMO don't oppose, but in the absence of tests (not your fault, of course, I only now realised I didn't add any) I think it should happen incrementally.

Can we identify the exact differences between 1 and 2 grammars, and do those individually? I don't know if there's a changelog per se, I'm only aware of https://aws.amazon.com/blogs/developer/introducing-smithy-idl-2-0/

From the ones I know, it's the 

1. Inline structures
2. Inline input/output params
3. enum and intEnum quiet syntax?

I don't think those warrant a complete rewrite of the grammar (for now).

With regards to the C file - I'm not sure what nvim-treesitter will do if it doesn't see the `parser.c` in the branch - I really hope that it will regenerate the grammar. If that's the case, it's excellent and is much easier to review and safer in general.

**Comment by benoitlouy** [comment]
Hi @keynmol, in my last commit I updated the grammar to be an evolution of what is in the main branch. This should make reviewing the PR easier.
I will try to leave some comments in the PR to point out the changes.

File: grammar.js [hunk 1]
```diff
@@ -10,27 +11,30 @@ module.exports = grammar({
     // TODO: add the actual grammar rules
     source_file: $ =>
       seq(
-        $.preamble,
-        repeat($._definition)
+        optional($.control_section),
+        optional($.metadata_section),
+        optional($.shape_section)
```
Comment (benoitlouy) [inline grammar.js:16]: I reorganized the general structure of a source file so it's closer to [`idl` definition from the doc](https://awslabs.github.io/smithy/2.0/spec/idl.html#smithy-idl-abnf)

File: grammar.js [hunk 2]
```diff
@@ -10,27 +11,30 @@ module.exports = grammar({
...
 
+    shape_section: $ => seq(
+      $.namespace_statement,
+      repeat($._definition)
+    ),
+
     _definition: $ => choice(
       $.use_statement,
-      seq($.trait_statements, $.shape_statement),
+      $.apply_trait_statement,
```
Comment (benoitlouy) [inline grammar.js:26]: `apply` statements weren't part of the grammar. They are now correctly recognized.

File: grammar.js [hunk 3]
```diff
@@ -10,27 +11,30 @@ module.exports = grammar({
...
       ),
 
+    shape_section: $ => seq(
+      $.namespace_statement,
+      repeat($._definition)
+    ),
+
     _definition: $ => choice(
       $.use_statement,
-      seq($.trait_statements, $.shape_statement),
```
Comment (benoitlouy) [inline grammar.js:19]: I removed this choice and updated `shape_statement` to include `optional($.trait_statements)`.

File: grammar.js [hunk 4]
```diff
@@ -39,72 +43,111 @@ module.exports = grammar({
...
+      $.shape_members
+    ),
+    union_shape_statement: $ => seq('union', $.identifier, optional($.mixins), $.shape_members),
 
-    service_shape_statement: $ => seq('service', $.identifier, $.node_object),
-    operation_shape_statement: $ => seq('operation', $.identifier, $.node_object),
-    resource_shape_statement: $ => seq('resource', $.identifier, $.node_object),
+    service_shape_statement: $ => seq('service', $.identifier, optional($.mixins), $.node_object),
+    operation_shape_statement: $ => seq('operation', $.identifier, optional($.mixins), $.operation_body),
+    resource_shape_statement: $ => seq('resource', $.identifier, optional($.mixins), $.node_object),
```
Comment (benoitlouy) [inline grammar.js:72]: - Add support for the mixins syntax
- Add support for the structure resource syntax
- Update the operation shape body to support inline structure definitions.

File: grammar.js [hunk 5]
```diff
@@ -39,72 +43,111 @@ module.exports = grammar({
...
+      optional($.structure_resource),
+      optional($.mixins),
+      $.shape_members
     ),
 
     trait_statements: $ => repeat1($.trait),
-    trait: $ => seq('@', $.shape_id, repeat($.trait_body)),
-    trait_body: $ => seq('(', $.trait_body_value, ')'),
+    trait: $ => seq('@', $.shape_id, optional($.trait_body)),
+    trait_body: $ => seq('(', optional($.trait_body_value), ')'),
```
Comment (benoitlouy) [inline grammar.js:118]: `@mytrait()` is a valid declaration. See https://awslabs.github.io/smithy/2.0/spec/idl.html#grammar-token-smithy-TraitBody

File: grammar.js [hunk 6]
```diff
@@ -39,72 +43,111 @@ module.exports = grammar({
...
+      optional($.trait_statements),
+      optional($.structure_resource),
+      optional($.mixins),
+      $.shape_members
     ),
 
     trait_statements: $ => repeat1($.trait),
-    trait: $ => seq('@', $.shape_id, repeat($.trait_body)),
-    trait_body: $ => seq('(', $.trait_body_value, ')'),
+    trait: $ => seq('@', $.shape_id, optional($.trait_body)),
```
Comment (benoitlouy) [inline grammar.js:117]: The use of `repeat` looks to be a mistake here and should be `optional` instead. See https://awslabs.github.io/smithy/2.0/spec/idl.html#grammar-token-smithy-Trait

File: grammar.js [hunk 7]
```diff
@@ -39,72 +43,111 @@ module.exports = grammar({
...
+    trait_body: $ => seq('(', optional($.trait_body_value), ')'),
     trait_body_value: $ => choice($.trait_structure, $.node_value),
-    trait_structure: $ => seq($.trait_structure_kvp, repeat(seq(',', $.trait_structure_kvp))),
+    trait_structure: $ => repeat1($.trait_structure_kvp),
```
Comment (benoitlouy) [inline grammar.js:120]: Because commas are now optional, the previous rule can be simplified by using `repeat1`

File: grammar.js [hunk 8]
```diff
@@ -39,72 +43,111 @@ module.exports = grammar({
...
     trait_structure_kvp: $ => seq($.node_object_key, ':', $.node_value),
-    apply_trait_statement: $ => seq('apply', $.shape_id, $.trait),
+
+    apply_trait_statement: $ => choice($.apply_statement_singular, $.apply_statement_block),
+    apply_statement_singular: $ => seq('apply', $.shape_id, $.trait),
+    apply_statement_block: $ => seq('apply', $.shape_id, '{', optional($.trait_statements), '}'),
```
Comment (benoitlouy) [inline grammar.js:125]: Updated the `apply` trait rules to support the block syntax.
```kotlin
apply Data {
  @trait1
  @trait2(foo: "bar")
}

File: grammar.js [hunk 9]
```diff
@@ -39,72 +43,111 @@ module.exports = grammar({
...
+      '[',
+      repeat($.node_value),
+      ']'
+    ),
+    node_object: $ => seq(
       '{',
-      $.node_object_kvp,
-      repeat(seq(',', $.node_object_kvp)),
-      optional(','),
+      repeat($.node_object_kvp),
```
Comment (benoitlouy) [inline grammar.js:144]: Simplifications because commas are now optional.

File: grammar.js [hunk 10]
```diff
@@ -39,72 +43,111 @@ module.exports = grammar({
...
       '}'
     ),
     node_object_kvp: $ => seq($.node_object_key, ':', $.node_value),
     node_keywords: $ => /true|false|null/,
 
-    identifier: $ => /([_]*)([a-zA-Z])([a-zA-Z0-9_]+)/,
+    identifier: $ => /([_]*)([a-zA-Z])([a-zA-Z0-9_]*)/,
```
Comment (benoitlouy) [inline grammar.js:150]: Allow single char identifiers.

**Comment by benoitlouy** [comment]
Hi @keynmol, is there anything I can do to help with this PR?

**Comment by keynmol** [comment]
@benoitlouy I think the PR looks good, it's entirely my fault that I completely forgot to review it :) 

I will do it as soon as possible, hopefully tomorrow.

File: grammar.js (hunk 7 above)
Comment (keynmol) [reply grammar.js:120]: Note to myself - you added comma to extras, that's why it all works. Nice.

**Review by keynmol — APPROVED** [review]
Really good work!

I ran it on a few specs and it seems to be performing well on the new syntax 🎉 
Also really liked the restructuring - commas being optional makes a lot of things much nicer and easier.

**Comment by keynmol** [comment]
Btw, I don't know which editor you use, but if it's anything that allows you to do LSP easily, I recommend my other project which is a LSP for grammar.js files :D https://github.com/keynmol/grammar-js-lsp

**Pull request merged**